mix/exit_layer.nim

```diff
@@ -0,0 +1,104 @@
+import std/[enumerate, strutils]
+import chronicles, chronos, metrics
+import libp2p, libp2p/[builders, stream/connection]
+import ./[mix_metrics, exit_connection, serialization, utils]
+
+type ProtocolHandler* = proc(conn: Connection, codec: string): Future[void] {.
+  async: (raises: [CancelledError])
+.}
+
+type ExitLayer* = object
+  switch: Switch
+  pHandler: ProtocolHandler
+
+proc callHandler(
+    switch: Switch, conn: Connection, codec: string
+): Future[void] {.async: (raises: [CatchableError]).} =
+  for index, handler in enumerate(switch.ms.handlers):
+    if codec in handler.protos:
+      await handler.protocol.handler(conn, codec)
+      return
+  error "Handler doesn't exist", codec = codec
+
+proc init*(T: typedesc[ExitLayer], switch: Switch): T =
+  ExitLayer(
+    switch: switch,
+    pHandler: proc(
+        conn: Connection, codec: string
+    ): Future[void] {.async: (raises: [CancelledError]).} =
+      try:
+        await callHandler(switch, conn, codec)
+      except CatchableError as e:
+        error "Error during execution of MixProtocol handler: ", err = e.msg
+    ,
+  )
+
+proc runHandler(
+    self: ExitLayer, codec: string, message: seq[byte]
+) {.async: (raises: [CancelledError]).} =
+  trace "Received: ", receiver = multiAddr, codec, message
+
+  let exitConn = MixExitConnection.new(message)
+
+  await self.pHandler(exitConn, codec)
+
+  if exitConn != nil:
+    try:
+      await exitConn.close()
+    except CatchableError as e:
+      error "Failed to close exit connection: ", err = e.msg
+    return
```

Comment:
use option/result instead of nil. If this is out of scope for this PR, comment/create GH issue as appropriate.

Reply:
Fixed in https://github.com/vacp2p/mix/pull/62